### ISSUE TITLE
Plex.Request.LibraryItems for .photo type

### ISSUE THREAD

**Issue by Jue-Xu** [lcharlick/PlexKit#13]
Hi, thank you very much for your work for PlexKit which helps me a lot! In my project, I want to list the items in my photo library with the code
```
        let info = Plex.ClientInfo(clientIdentifier: UUID().uuidString)
        let client = Plex(sessionConfiguration: .default, clientInfo: info)
        
        client.request(Plex.Request.LibraryItems(key: "11", mediaType: .photo),
                       from: url,
                       token: plexToken
        ) { result in
            switch result {
            case .success(let response):
                print(response)
            case .failure(let error):
                print("An error occurred: \(error)")
            }
        }
```
but I get nil metadata as 
```
Response(mediaContainer: PlexKit.Plex.Request._LibraryItems<PlexKit.PlexMediaItem>.Response.MediaContainer(size: 0, totalSize: nil, allowSync: Optional(true), art: Optional("/:/resources/photo-fanart.jpg"), identifier: Optional("com.plexapp.plugins.library"), librarySectionID: Optional(11), librarySectionTitle: Optional("Photos"), librarySectionUUID: Optional("08f24eb6-3851-4fa3-ae42-de20ff8accf8"), mediaTagPrefix: Optional("/system/bundle/media/flags/"), mediaTagVersion: Optional(1649089686), nocache: Optional(true), offset: nil, thumb: Optional("/:/resources/photo.png"), title1: Optional("Photos"), title2: Optional("All Photos"), viewGroup: Optional(PlexKit.PlexMediaType.photo), viewMode: Optional(131633), Metadata: nil))
```
I tried similar requests for .show, .movie or .artist. All of them worked. Only .photo type didn't return the expected result as the result returned by
```
http://url:32400/library/sections/11/all?X-Plex-Token=mytoken
```

**Comment by lcharlick** [comment]
Hi,

Can you post the JSON response for that URL? Perhaps the shape of the `Metadata` object is different for photo libraries.

**Comment by Jue-Xu** [comment]
OMG! Thanks for you so quick reply! 

```
<MediaContainer size="16" allowSync="1" art="/:/resources/photo-fanart.jpg" identifier="com.plexapp.plugins.library" librarySectionID="11" librarySectionTitle="Photos" librarySectionUUID="08f24eb6-3851-4fa3-ae42-de20ff8accf8" mediaTagPrefix="/system/bundle/media/flags/" mediaTagVersion="1649089686" nocache="1" thumb="/:/resources/photo.png" title1="Photos" title2="All Photos" viewGroup="photo" viewMode="131633">
<Directory ratingKey="70037" key="/library/metadata/70037/children" guid="local://70037" type="photo" title="「空をこえて」" summary="" index="1" composite="/library/metadata/70037/composite/1648015424" thumb="/library/metadata/70037/thumb/1648015424" art="/library/metadata/70037/art/1648015424" addedAt="1642919816" updatedAt="1648015424"> </Directory>
...
<Directory ratingKey="58672" key="/library/metadata/58672/children" guid="local://58672" type="photo" title="本田翼" summary="" index="1" composite="/library/metadata/58672/composite/1647960308" thumb="/library/metadata/58672/thumb/1647960308" art="/library/metadata/58672/art/1647960308" addedAt="1642692226" updatedAt="1647960308"> </Directory>
...
</MediaContainer>
```
For simplicity, I omitted most `<Directory> ...</Directory>`.

**Comment by lcharlick** [comment]
The Plex API returns XML by default, unfortunately. 
If you could include `Accept: application/json` in the request header that'll force it to return JSON instead (which is what PlexKit uses).

I should set up a photo library myself, but I won't have a chance to get to it today - hoping there's a quick work around we can use in the meantime 🙂

**Comment by Jue-Xu** [comment]
Thanks again for your help! I am not sure if it is what you expect
```
HTTP/1.1 200 OK
X-Plex-Protocol: 1.0
Content-Type: application/json
Content-Length: 6015
Connection: close
Cache-Control: no-cache
Date: Tue, 10 May 2022 13:04:55 GMT

{"MediaContainer":{"size":16,"allowSync":true,"art":"/:/resources/photo-fanart.jpg","identifier":"com.plexapp.plugins.library","librarySectionID":11,"librarySectionTitle":"Photos","librarySectionUUID":"08f24eb6-3851-4fa3-ae42-de20ff8accf8","mediaTagPrefix":"/system/bundle/media/flags/","mediaTagVersion":1649089686,"nocache":true,"thumb":"/:/resources/photo.png","title1":"Photos","title2":"All Photos","viewGroup":"photo","viewMode":131633,
"Metadata":[{"ratingKey":"70037","key":"/library/metadata/70037/children","guid":"local://70037","type":"photo","title":"「空をこえて」","summary":"","index":1,"composite":"/library/metadata/70037/composite/1648015424","thumb":"/library/metadata/70037/thumb/1648015424","art":"/library/metadata/70037/art/1648015424","addedAt":1642919816,"updatedAt":1648015424},
...
{"ratingKey":"58672","key":"/library/metadata/58672/children","guid":"local://58672","type":"photo","title":"本田翼","summary":"","index":1,"composite":"/library/metadata/58672/composite/1647960308","thumb":"/library/metadata/58672/thumb/1647960308","art":"/library/metadata/58672/art/1647960308","addedAt":1642692226,"updatedAt":1647960308}]}}
```
<img width="488" alt="image" src="https://user-images.githubusercontent.com/16731479/167636251-78bb85e9-4743-497f-86b9-982cae2873c9.png">

<img width="473" alt="image" src="https://user-images.githubusercontent.com/16731479/167636144-a04d8901-2ca8-4d1d-b5cb-b26166fe97e4.png">

Look forward to your reply!

**Comment by lcharlick** [comment]
Strange - that all parses fine on my end. I also set up a basic photo library and copied the request from your first post, and I get a non-nil metadata response:

<img width="560" alt="Screen Shot 2022-05-10 at 11 16 06 pm" src="https://user-images.githubusercontent.com/783664/167643832-bfd79829-c129-427e-9ddf-d5ae403e44d3.png">

**Comment by Jue-Xu** [comment]
Thanks again! It looks that the problem is on my side. I will take some time to read your source code and Plex API to find my problem. I will let you know if I solve the issue or I am still stuck.

FYI, here are two screenshots of my console for two cases
nil metadata for .photo
<img width="836" alt="image" src="https://user-images.githubusercontent.com/16731479/167646423-39f37b84-f5ab-4b2e-9b84-8090da81d522.png">
but works for .show
<img width="845" alt="image" src="https://user-images.githubusercontent.com/16731479/167647496-7faa3529-0ef1-4c91-86fa-7ea61502eb73.png">

**Comment by Jue-Xu** [comment]
OK, I somewhat find the problem. I create a new simple photo library which contains several photos and one folder. The same code in my first post returns non-nil metadata 
<img width="824" alt="image" src="https://user-images.githubusercontent.com/16731479/167651783-afd27b00-3aa1-4a20-8a65-b81ab0a4a482.png">

I find the difference is that the new photo library returns 
![image](https://user-images.githubusercontent.com/16731479/167652793-d0d65668-3ff6-4a30-8f23-fdf262aeb1cb.png)
in which the items are `<Photo>`. That means only `<Photo>` is parsed. While in my first example, all items are `< Directory >` which are all folders. So, I tried adding a single photo to my first photo library, the code in my first post can return the metadata of that single photo.

**Comment by lcharlick** [comment]
Ah, I see. The `/library/sections/<library>/all` endpoint doesn't return photos in nested directories. The web UI uses `/library/sections/<library>/folder` instead. 

I'll look into adding support for that endpoint (give me a day or two).

**Comment by lcharlick** [comment]
@Jue-Xu give the `photo-libraries` (PR #14) branch a go when you get a chance. I've added a new `Folder` request type there. Here's a quick and dirty example:

```swift
client.request(Plex.Request.Folder(key: libraryKey, mediaType: .photo, parentRatingKey: nil), from: server, token: token) { result in
    switch result {
    case .success(let response):
        switch response.mediaContainer.metadata[0] {
        case .folder(let folder):
            print("First item is a folder (\(folder.title))")
            client.request(Plex.Request.Folder(key: libraryKey, mediaType: .photo, parentRatingKey: folder.ratingKey), from: server, token: token) { result in
                switch result {
                case .success(let nestedResponse):
                    switch nestedResponse.mediaContainer.metadata[0] {
                    case .folder(let folder):
                        print("Found a nested folder: \(folder)")
                    case .mediaItem(let nestedItem):
                        print("Found a nested media item (\(nestedItem.title!))")
                    }
                case .failure(let nestedError):
                    print("An error occurred: \(nestedError)")
                }
            }
        case .mediaItem:
            print("First item is a media item!")
        }
    case .failure(let error):
        print("An error occurred: \(error)")
    }
}
```

**Comment by Jue-Xu** [comment]
Thanks again for your help!

I am not sure whether I am on the right way to use the branch (I am new to iOS development): I downloaded `photo-libraries` branch and replace the (default) PlexKit folder with it on my machine. I think I have already got the latest update because I saw 
<img width="1174" alt="image" src="https://user-images.githubusercontent.com/16731479/167806710-c60a3581-9d1b-41d0-9169-0d86167ec1eb.png">

However, I got the errors when I tried to compile
<img width="1167" alt="image" src="https://user-images.githubusercontent.com/16731479/167804585-13e7cc0b-906c-4f73-ac88-a303af0e10ea.png">

**Comment by lcharlick** [comment]
My mistake - try now. 
P.S. You can point Xcode at a specific package branch in project settings - no need to manually download.

<img width="760" alt="Screen Shot 2022-05-11 at 6 19 32 pm" src="https://user-images.githubusercontent.com/783664/167810122-8400ea19-36c7-489e-a485-f3c27eb26e5f.png">

**Comment by Jue-Xu** [comment]
It works! I have the response 

```
First item is a folder (「空をこえて」)
Found a nested media item (6b63e9b7gy1fqfh92mux2j215r1o0x2l)
```

**Comment by lcharlick** [comment]
Great! I'll clean up that PR a bit then create a new release (1.6.0 probably).

**Comment by lcharlick** [comment]
Fixed in v1.6.0

**Comment by Jue-Xu** [comment]
Recently, I managed to target the reason why photo folders are not parsed with `Plex.Request.LibraryItems(key: plexPhotoLibraryKey, mediaType: .photo)`: photo has type key=13, while photo folder has type key=14. 

For example, my photo library contains one image and seven photo folders
<img width="1056" alt="image" src="https://github.com/lcharlick/PlexKit/assets/16731479/77601239-6381-4528-868e-94d1922c1cb2">

<img width="1034" alt="image" src="https://github.com/lcharlick/PlexKit/assets/16731479/d7861082-ef14-4442-bbf4-61b4e98ac98e">

<img width="1029" alt="image" src="https://github.com/lcharlick/PlexKit/assets/16731479/96cae4e6-c847-4461-bc09-050d650c6163">


So, we could solve this issue by adding a new PlexMediaType e.g., photoFolder with key=14, and update PlexMediaType.swift correspondingly as follows. Would you mind updating it to your PlexKit package?

```swift
//
//  PlexMediaType.swift
//  PlexKit
//
//  Created by Lachlan Charlick on 27/5/20.
//  Copyright © 2020 Lachlan Charlick. All rights reserved.
//

import Foundation

/// - Note: https://github.com/Arcanemagus/plex-api/wiki/MediaTypes
public enum PlexMediaType: Hashable {
    case unknown(String)

    case movie
    case show
    case season
    case episode
    case trailer
    case comic
    case person

    case artist
    case album
    case track
    case photoAlbum
    case picture
    case photo
    case photoFolder

    case playlist
    case playlistFolder
    case userPlaylistItem

    case collection

    init(rawValue: String) {
        switch RawValue(rawValue: rawValue) {
        case .movie:
            self = .movie
        case .show:
            self = .show
        case .season:
            self = .season
        case .episode:
            self = .episode
        case .trailer:
            self = .trailer
        case .comic:
            self = .comic
        case .person:
            self = .person
        case .artist:
            self = .artist
        case .album:
            self = .album
        case .track:
            self = .track
        case .photoAlbum:
            self = .photoAlbum
        case .picture:
            self = .picture
        case .photo:
            self = .photo
        case .photoFolder:
            self = .photoFolder
        case .playlist:
            self = .playlist
        case .playlistFolder:
            self = .playlistFolder
        case .userPlaylistItem:
            self = .userPlaylistItem
        case .collection:
            self = .collection
        default:
            self = .unknown(rawValue)
        }
    }

    public var rawValue: String {
        switch self {
        case let .unknown(rawValue):
            return rawValue
        case .movie:
            return RawValue.movie.rawValue
        case .show:
            return RawValue.show.rawValue
        case .season:
            return RawValue.season.rawValue
        case .episode:
            return RawValue.episode.rawValue
        case .trailer:
            return RawValue.trailer.rawValue
        case .comic:
            return RawValue.comic.rawValue
        case .person:
            return RawValue.person.rawValue
        case .artist:
            return RawValue.artist.rawValue
        case .album:
            return RawValue.album.rawValue
        case .track:
            return RawValue.track.rawValue
        case .photoAlbum:
            return RawValue.photoAlbum.rawValue
        case .picture:
            return RawValue.picture.rawValue
        case .photo:
            return RawValue.photo.rawValue
        case .photoFolder:
            return RawValue.photoFolder.rawValue
        case .playlist:
            return RawValue.playlist.rawValue
        case .playlistFolder:
            return RawValue.playlistFolder.rawValue
        case .userPlaylistItem:
            return RawValue.userPlaylistItem.rawValue
        case .collection:
            return RawValue.collection.rawValue
        }
    }

    enum RawValue: String, Codable, CaseIterable {
        case movie
        case show
        case season
        case episode
        case trailer
        case comic
        case person

        case artist
        case album
        case track
        case photoAlbum
        case picture
        case photo
        case photoFolder

        case playlist
        case playlistFolder
        case userPlaylistItem

        case collection
    }
}

extension PlexMediaType: Codable {
    public init(from decoder: Decoder) throws {
        let rawValue = try decoder.singleValueContainer().decode(String.self)
        self.init(rawValue: rawValue)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }
}

extension PlexMediaType {
    var key: Int {
        switch self {
        case .unknown:
            return -1

        case .movie:
            return 1
        case .show:
            return 2
        case .season:
            return 3
        case .episode:
            return 4
        case .trailer:
            return 5
        case .comic:
            return 6
        case .person:
            return 7
        case .artist:
            return 8
        case .album:
            return 9
        case .track:
            return 10
        case .photoAlbum:
            return 11
        case .picture:
            return 12
        case .photo:
            return 13
        case .photoFolder:
            return 14
        case .playlist:
            return 15
        case .playlistFolder:
            return 16
        case .collection:
            return 18
        case .userPlaylistItem:
            return 1001
        }
    }
}

```
